firmware/user/display/oled.c

```diff
@@ -202,6 +202,213 @@ void drawPixel(int16_t x, int16_t y, color c)
+        {
+            yerrnumerator = ((dy * sdy) * FIXEDPOINT + yerrdiv/2) / yerrdiv;
+            if( dxA )
+            {
+                int yn = (yerrnumerator * dxA)*sdy;
+                yn += FIXEDPOINTD2 * sdy; //This "feels" right
+                yn /= FIXEDPOINT;
+                y += yn;
+                //Weird situation - if we cal, and now, both ends are out on the same side abort.
+                if( y < 0 && y1 < 0 ) return;
+                if( y > BRESEN_H-1 && y1 > BRESEN_H-1 ) return;
+            }
+        }
+    }
+
+    if( xerrdiv > 0 )
+    {
+        int dyA = 0;    
+        if( y < 0 )
+        {
+            dyA = 0 - y;
+            y = 0;
+        }
+        if( y > BRESEN_H-1 )
+        {
+            dyA = (y - (BRESEN_H-1));
+            y = BRESEN_H-1;
+        }
+        if( dyA || xerrdiv > yerrdiv )
+        {
+            xerrnumerator = ((dx * sdx) * FIXEDPOINT + xerrdiv/2 ) / xerrdiv;
+            if( dyA )
+            {
+                int xn = (xerrnumerator*dyA)*sdx;
+                xn += FIXEDPOINTD2 * sdx; //This "feels" right.
+                xn /= FIXEDPOINT;
+                x += xn;
+                //If we've come to discover the line is actually out of bounds, abort.
+                if( x < 0 && x1 < 0 ) return;
+                if( x > BRESEN_W-1 && x1 > BRESEN_W-1 ) return;
+            }
+        }
+    }
+
+    if( x1 == x && y1 == y )
+    {
+        drawPixelFastWhite( x, y );
+        return;
+    }
+
+    //Make sure we haven't clamped the wrong way.
+    //Also this checks for vertical/horizontal violations.
+    if( dx > 0 )
+    {
+        if( x > BRESEN_W-1 ) return;
+        if( x > x1 ) return;
+    }
+    else if( dx < 0 )
+    {
+        if( x < 0 ) return;
+        if( x < x1 ) return;
+    }
+
+    if( dy > 0 )
+    {
+        if( y > BRESEN_H-1 ) return;
+        if( y > y1 ) return;
+    }
+    else if( dy < 0 )
+    {
+        if( y < 0 ) return;
+        if( y < y1 ) return;
+    }
+
+    //Force clip end coordinate.
+    //NOTE: We have an extra check within the inner loop, to avoid complicated math here.
+    //Theoretically, we could math this so that in the end-coordinate clip stage
+    //to make sure this condition just could never be hit, however, that is very
+    //difficult to guarantee under all situations and may have weird edge cases.
+    //So, I've decided to stick this here.
+
+    if( xerrdiv > yerrdiv )
+    {
+        int xerr = FIXEDPOINTD2;
+        if( x1 < 0 ) x1 = 0;
+        if( x1 > BRESEN_W-1) x1 = BRESEN_W-1;
+        x1 += sdx; //Tricky - make sure the "next" mark we hit doesn't overflow.
+
+        if( y1 < 0 ) y1 = 0;
+        if( y1 > BRESEN_H-1 ) y1 = BRESEN_H-1;
+
+        for( ; y != y1; y+=sdy )
+        {
+            drawPixelFastWhite( x, y );
+            xerr += xerrnumerator;
+            while( xerr >= FIXEDPOINT )
+            {
+                x += sdx;
+                if( x == x1 ) goto abortline;
```

Comment:
This is functionally a return, unless you were planning to do something at that label.

Reply:
Correct